development/extensions/tutorial_controllers.rst

```diff
@@ -0,0 +1,285 @@
+
+In case the name is not ``bertie`` use the ``phpbb\controller\helper`` to render
+a style template file called ``demo_body.html``.
+
+Template file
+=============
+
+The template file has to be in the ``styles/prosilver/template/`` directory,
+similar to the listener file. So we create the HTML file ``demo_body.html`` in
+that directory using the following content including the phpBB header and
+footer:
+
+.. code-block:: html
+
+    <!-- INCLUDE overall_header.html -->
+
+    <h2>{DEMO_MESSAGE}</h2>
+
+    <!-- INCLUDE overall_footer.html -->
+
+Routing
+=======
+
+Now the controller would display the page content, but we don't have a URL which
+executes the controller. The URLs are defined in the ``config/routing.yml`` file
+of our extension:
+
+.. code-block:: yaml
+
+    acme_demo_route:
+        path: /demo/{name}
+        defaults: { _controller: acme.demo.controller:handle }
+
+
+.. warning::
+
+    Similar to event and service names your route names should be prefixed with
+    your vendor and extension name.
+
+``path`` specifies the URL component. Curly braces are used as placeholders. The
+name of the placeholder is the name of the variable of the controller's
+``handle()`` method. The order of the arguments may also be different in the URL
+and method.
+
+In the ``defaults`` section the service name of the controller is specified. In
+the `Dependencies`_ section we used ``acme.demo.controller`` as a service name.
+``handle`` is the name of the method that should be called on the controller.
+
+You can also specify a regular expression, to make sure that your controller is
+only used, when a parameter matches it. For example, if we want to ensure that
+``name`` is an integer, we could add the following code to the route definition:
+
+.. code-block:: yaml
+
+        requirements:
+            page: \d+
+
+Now if you type the link ``https://localhost/phpBB/app.php/demo/world`` into
+your browser, you will see the page.
+
+Generating links to routes
+==========================
+
+In order to populate the ``a``-element in navigation with our link, we need to
+assign the link to the template variable ``U_DEMO_PAGE``.
+
+We have to use the ``core.page_header`` event for that. So we replace our old
+``getSubscribedEvents()`` method in the ``event/main_listener.php`` and also add
+the template engine and controller helper as dependencies:
+
+.. code-block:: php
+
+        static public function getSubscribedEvents()
+        {
+            return array(
+                'core.user_setup'  => 'load_language_on_setup',
+                'core.page_header' => 'add_page_header_link',
+            );
+        }
+
+        /* @var helper */
+        protected $helper;
+
+        /* @var template */
+        protected $template;
+
+        /**
+         * Constructor
+         *
+         * @param helper $helper
+         * @param template $template
+         */
+        public function __construct(helper $helper, template $template)
```

Comment:
`helper` is the type hinting. There is no need to add the namespace, when the class is imported with a `use` statement in the beginning of the php file.